controller_server:
  ros__parameters:
    controller_frequency: 30.0
    costmap_update_timeout: 0.5
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.001
    min_theta_velocity_threshold: 0.001
    failure_tolerance: 0.5
    progress_checker_plugins: ["progress_checker"]
    goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
    controller_plugins: ["FollowPath"]
    use_realtime_priority: false
    speed_limit_topic: "speed_limit"

    # Progress checker parameters
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.5
      movement_time_allowance: 30.0
    # Goal checker parameters
    #precise_goal_checker:
    #  plugin: "nav2_controller::SimpleGoalChecker"
    #  xy_goal_tolerance: 0.25
    #  yaw_goal_tolerance: 0.25
    #  stateful: True
    general_goal_checker:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.1
      yaw_goal_tolerance: 0.1
    FollowPath:
      plugin: "nav2_mppi_controller::MPPIController"
      time_steps: 60
      model_dt: 0.05
      batch_size: 1000
      ax_max: 1.5
      ax_min: -1.5
      ay_max: 1.0
      ay_min: -1.0
      az_max: 2.0
      vx_std: 0.8
      vy_std: 0.8
      wz_std: 0.5
      vx_max: 1.0
      vx_min: -0.8
      vy_max: 1.0
      vy_min: -1.0
      wz_max: 1.5
      iteration_count: 1
      prune_distance: 2.5 # last_vlaue: 5.0
      transform_tolerance: 1.0
      temperature: 0.5 # last_value: 0.3 
      gamma: 0.015
      motion_model: "Omni"
      visualize: true
      publish_optimal_trajectory: true
      retry_attempt_limit: 5
      regenerate_noises: true
      TrajectoryVisualizer:
        trajectory_step: 5
        time_step: 3
      TrajectoryValidator:
        # The validator can be configured with additional parameters if needed.
        plugin: "mppi::DefaultOptimalTrajectoryValidator"
        collision_lookahead_time: 2.0  # Time to look ahead to check for collisions
        consider_footprint: true  # Whether to consider the full robot's footprint (or circle) in trajectory validation collision checks
      AckermannConstraints:
        min_turning_r: 0.2
      critics:
        [
          "ConstraintCritic",
          "CostCritic",
          "GoalCritic",
          "GoalAngleCritic",
          "PathAlignCritic",
          "PathFollowCritic",
          "PathAngleCritic",
          # "PreferForwardCritic",
        ]
      ConstraintCritic:
        enabled: true
        cost_power: 1
        cost_weight: 4.0
      GoalCritic:
        enabled: true
        cost_power: 1
        cost_weight: 5.0
        threshold_to_consider: 1.4
      GoalAngleCritic:
        enabled: true
        cost_power: 1
        cost_weight: 3.0
        threshold_to_consider: 0.5
      PreferForwardCritic:
        enabled: true
        cost_power: 1
        cost_weight: 5.0
        threshold_to_consider: 0.5
      CostCritic:
        enabled: true
        cost_power: 1
        cost_weight: 3.81
        near_collision_cost: 253
        critical_cost: 300.0
        consider_footprint: true
        collision_cost: 1000000.0
        near_goal_distance: 1.0
        trajectory_point_step: 2
      PathAlignCritic:
        enabled: true
        cost_power: 1
        cost_weight: 8.0 #last_value: 14.0
        max_path_occupancy_ratio: 0.05
        trajectory_point_step: 4
        threshold_to_consider: 0.5
        offset_from_furthest: 20
        use_path_orientations: true
      PathFollowCritic:
        enabled: true
        cost_power: 1
        cost_weight: 8.0 #last_value: 5.0
        offset_from_furthest: 5
        threshold_to_consider: 1.4
      PathAngleCritic:
        enabled: true
        cost_power: 1
        cost_weight: 2.0
        offset_from_furthest: 4
        threshold_to_consider: 0.5
        max_angle_to_furthest: 1.0
        mode: 0
      # TwirlingCritic:
      #   enabled: true
      #   twirling_cost_power: 1
      #   twirling_cost_weight: 10.0

local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 30.0
      publish_frequency: 30.0
      global_frame: odom
      robot_base_frame: base_link
      rolling_window: true
      width: 20
      height: 20
      resolution: 0.05
      # robot_radius: 0.0
      footprint: "[[0.305, 0.185], [0.305, -0.185], [-0.305, -0.185], [-0.305, 0.185]]"
      footprint_padding: 0.05
      plugins: ["nonpersisting_obstacle_layer", "inflation_layer"]
      transform_tolerance: 1.0
      initial_transform_timeout: 60.0

      voxel_layer:
        plugin: "nav2_costmap_2d::VoxelLayer"
        enabled: True
        publish_voxel_map: True        # 调试时可先 True，看 voxel 地图
        origin_z: 0.0
        z_resolution: 0.05
        z_voxels: 16
        max_obstacle_height: 2.0        # 机器人以上 2m 内算障碍
        mark_threshold: 0

        observation_sources: livox_points
        livox_points:
          topic: /livox/lidar           # 改成你的 Mid-360 点云话题
          data_type: "PointCloud2"      # 关键：用 PointCloud2
          clearing: True
          marking: True
          obstacle_max_range: 12.0      # 有效避障距离（建议 10~15m）
          obstacle_min_range: 0.2       # 略大于 0.1m，过滤近距离噪点
          raytrace_max_range: 15.0      # 清障射线最大距离
          raytrace_min_range: 0.2
          min_obstacle_height: 0.1      # 地面以上 10cm 认为是障碍
          max_obstacle_height: 2.0

      nonpersisting_obstacle_layer:
            plugin: nav2_costmap_2d/NonPersistentVoxelLayer
            enabled:              true
            track_unknown_space:  true
            max_obstacle_height:  2.0
            unknown_threshold:    15
            mark_threshold:       2
            combination_method:   1
            obstacle_range: 15.0
            origin_z: 0.0
            z_resolution: 0.05
            z_voxels: 16
            publish_voxel_map: true
            observation_sources: livox_points
            livox_points:
              topic: /livox/lidar           # 改成你的 Mid-360 点云话题
              data_type: "PointCloud2"      # 关键：用 PointCloud2
              # clearing: True
              marking: True
              obstacle_max_range: 15.0      # 有效避障距离（建议 10~15m）
              obstacle_min_range: 0.2       # 略大于 0.1m，过滤近距离噪点
              raytrace_max_range: 15.0      # 清障射线最大距离
              raytrace_min_range: 0.2
              min_obstacle_height: 0.1      # 地面以上 10cm 认为是障碍
              max_obstacle_height: 2.0

      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 4.0
        inflation_radius: 0.4

